### DISCUSSION TITLE
What's the required steps to setup `monaco-languageclient`?

### DISCUSSION THREAD

**Discussion by Tnze** [TypeFox/monaco-languageclient#585] [Q&A]
After #583, I already established the connection between language server and this package.

But it doesn't seem to be working:
![image](https://github.com/TypeFox/monaco-languageclient/assets/19628575/4f24160d-2953-4797-86eb-ab9c55335663)

I think I didn't configure `monaco-languageclient` in the right way. But I don't know how to find the problem.

I suspect that's because I configured `monaco-editor` first, and introduce `monaco-languageclient` later. So I didn't call the right setup function. For example, I call `monaco.editor.create` instead of `createConfiguredEditor` from `vscode/monaco`, does it matter? I also tried to use `createConfiguredEditor` but it doesn't work too.

I also have highlight working before introducing `monaco-languageclient`, so it also maybe is the `pnpm` overrides doesn't work property?

The current situation leaves me with no clue. I would be very grateful if someone could provide any ideas.

**Comment by kaisalmen** [comment]
Hi @Tnze can you share your code or provide an example in a separate GitHub repository? Thank you

**Reply by Tnze** [reply]
Hi, I add the access for you for my private repo (It's not yet ready to public).
To run the project you may want to follow the [Tauri Guide](https://tauri.app/v1/guides/getting-started/prerequisites)

**Reply by kaisalmen** [reply]
Thank you, I received it. I won't be able to look it at deeply today, but in the upcoming days.

**Comment by Tnze** [comment]
I got the lsp functions working! By add this line:
```typescript
monaco.languages.register({ id: 'c', aliases: ['c'] })
```
![image](https://github.com/TypeFox/monaco-languageclient/assets/19628575/fb019728-c6ee-4750-9467-7666aef1cf71)

But highlight is still missing. I guess some override service are needed.

**Reply by Tnze** [reply]
Highlights doesn't work, seems to related #511. But I have no idea because I'm using vite.

**Reply by Tnze** [reply]
The [`monaco-vscode-api`](https://github.com/CodinGame/monaco-vscode-api) disabled the default highlight, see: https://github.com/CodinGame/monaco-vscode-api/issues/183#issuecomment-1708429486

To add it back:
```ts
import 'monaco-editor/esm/vs/basic-languages/monaco.contribution.js'
```

**Reply by kaisalmen** [reply]
@Tnze or you can do it like this. 
```js
import '@codingame/monaco-vscode-theme-defaults-default-extension';
import '@codingame/monaco-vscode-json-default-extension';
```
Didn't realize that was the problem. This is done in several of our examples, like here:
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/common/client-commons.ts#L8-L9

**Reply by kaisalmen** [reply]
These are separate package exports, see here:
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/package.json#L46-L57

**Reply by Tnze** [reply]
Thank you!
By the way, how could I use the Chinese language pack? I add 
```ts
import getLanguageServiceOverride from '@codingame/monaco-vscode-languages-service-override';
import '@codingame/monaco-vscode-language-pack-zh-hans';
```
and register the service in `initServices` but it doesn't work. Are there more steps to do?

**Reply by kaisalmen** [reply]
It must be loaded first. Look here:
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/statemachine_client.html
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/langium/localeLoader.ts

and you get the pack here:
https://www.npmjs.com/package/@codingame/monaco-vscode-language-pack-zh-hans

Btw, do you know @CGNonofr `monaco-vscode-api` demo?
https://codingame.github.io/monaco-vscode-api/

**Reply by Tnze** [reply]
Seems not work for me.

**Reply by kaisalmen** [reply]
Try if you get it working with a clone of this repo and just add the Chinese language pack in the example linked above.

**Reply by kaisalmen** [reply]
Don't forget to add it to the `package.json` update you installation.

**Reply by Tnze** [reply]
Yes, it works. I tried to add Chinese pack to the json example. It turns out that I have to load the language-pack package before `common/client-commons.ts`, otherwise it won't effect. I will test it on my project now.

**Reply by Tnze** [reply]
```typescript
import '@codingame/monaco-vscode-language-pack-zh-hans';
import * as monaco from 'monaco-editor';
```
The order of these two import statements is important. It's required to load language pack before loading monaco-editor.

**Reply by Tnze** [reply]
I don't think this is a good design. I believe modules should work in any import sequence.

**Reply by kaisalmen** [reply]
It is currently the way it is. Feel free to suggest an improvement: https://github.com/CodinGame/monaco-vscode-api

**Reply by Tnze** [reply]
I think I should import `@codingame/monaco-vscode-cpp-default-extension` for C language instead?
But I got this error:
![image](https://github.com/TypeFox/monaco-languageclient/assets/19628575/6a281bf5-90cf-4006-b2df-a42f9958a9d5)

It seems like a problem with `MemoryFileService2`, which also happens when I was trying to use `createConfiguredEditor` from `vscode/monaco`. How could I resolve this?

**Reply by kaisalmen** [reply]
This seems to be this problem: https://github.com/TypeFox/monaco-languageclient?tab=readme-ov-file#vite-dev-server-troubleshooting

**Reply by invm** [reply]
Hey @Tnze,

I'm working on something similar with Tauri and Monaco, I also want to remove the WS server and use Tauri IPC instead, can you share how you connected the Monaco frontend with an ls via Tauri IPC?

It'll help me a lot.

Thanks

**Reply by Tnze** [reply]
I provide a 'connect' command for the frontend. When calling this command, a lsp process is launched, and a numerical "session id" is returned. Two event are used after the session is created: `lsp-msg-in-{session_id}` and `lsp-msg-out-{session_id}`. Two tokio routine is spawned to copy the messages in and out. Another tokio routine is spawned and waiting the lsp process exit and send another event `lsp-close-{session_id}`.

**Reply by invm** [reply]
Thanks @Tnze, 

What about the frontend? Did you implement IConnectionProvider as you've mentioned in your previous discussion?
Do you manually emit lsp events?

**Reply by Tnze** [reply]
> Thanks @Tnze,
> 
> What about the frontend? Did you implement IConnectionProvider as you've mentioned in your previous discussion? Do you manually emit lsp events?

Yes, I extends AbstractMessageReader and implements MessageReader, extends AbstractMessageWriter and implements MessageWriter.

```ts
export async function createTauriMessageConnection(encoding: string) {
    if (encoding != 'utf8') throw "Unsupported encoding " + encoding
    const session = await invoke('connect') as number
    const reader = new TauriMessageReader(session)
    const writer = new TauriMessageWriter(session)
    return {
        reader: reader as MessageReader,
        writer: writer as MessageWriter,
    }
}
```

```ts
    languageClient = new MonacoLanguageClient({
        // .....
        connectionProvider: {
            get: createTauriMessageConnection
        }
    });
```

**Reply by invm** [reply]
Thank you, very much appreciated.
